io/src/main/java/org/commonjava/maven/ext/io/CharsetAwareJacksonJsonProvider.java

```diff
@@ -0,0 +1,83 @@
+/**
+ * Copyright (C) 2012 Red Hat, Inc. (jcasey@redhat.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.commonjava.maven.ext.io;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+
+import org.apache.commons.io.ByteOrderMark;
+import org.apache.commons.io.input.BOMInputStream;
+
+import com.fasterxml.jackson.core.JsonEncoding;
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.jayway.jsonpath.InvalidJsonException;
+import com.jayway.jsonpath.spi.json.JacksonJsonProvider;
+
+public class CharsetAwareJacksonJsonProvider extends JacksonJsonProvider
```

Comment:
Nice! Can we add some javadoc for the class description please?

Reply:
I was able to make the jpath library load both UTF-16 and UTF-32 files successfully by overriding the `parse()` method. Before this, `DocumentContext.jsonString()` in `JSONIO.writeJSON()` would just throw an exception.

The point wasn't even to necessarily support these encodings as I don't know of a specific case using it, but to check that the round-trip was correct after we load and save JSON files.

Currently, the round-trip tests still throw some exception since the `FileIO.determineEOL()` method only works on single-byte encodings.

For example, `'\n'` is nromally encoded as `'\n' = 0x0A`, but it is two bytes, `0x00 0x0A`, in a UTF-16BE file. Therefore, the current method of looking back a single byte will not work.

But, I don't know if it's worth spending time to fix this now unless we ever come across such a file in the wild.

Reply:
I think its already a nice improvement. I would agree thats its not worth spending time resolving this when we haven't encountered a problem ever so far.